src/components/structures/MatrixChat.tsx

```diff
@@ -756,6 +762,7 @@ export default class MatrixChat extends React.PureComponent<IProps, IState> {
                 Modal.createDialog(DialPadModal, {}, "mx_Dialog_dialPadWrapper");
                 break;
             case Action.OnLoggedIn:
+                this.stores.client = MatrixClientPeg.get();
```

Comment:
When this client is set like this, it won't cause already rendered components to notice as their `context` won't update due to `this.stores` not changing reference

Reply:
Then that should be in a comment on the context so its known and doesn't trip people up. I think it'd be nice if we handled the case of replacing values within the context properly, right now we only have immutables and MatrixClient which is set/replaced as per your comment but this context may include more values later down the line which don't quite fit that

Reply:
From my perspective, I would rather not over-engineer at this point. If and when we end up needing a truly mutable property on `Stores` we will then know the actual requirements for the actual feature we're trying to implement, as opposed to now where yes it would be nice to have but we don't know what API shape would work.

Reply:
We really ought to make this foolproof. Either by "over-engineering" or making use of typescript protections (e.g. readonly) or *plenty* of documentation.

Reply:
I'm happy to use Typescript `readonly` and docs for this. I feel trying to figure out an update and notification pattern at this point in time to be premature.

Reply:
> I feel trying to figure out an update and notification pattern at this point in time to be premature.

this would just be moving it to MatrixChat's state, and re-creating the object as

```
this.setState(state => ({
    sdkContext: {
        ...state.sdkContext,
        matrixClient: newClient,
    },
}));
```

and passing `this.state.sdkContext` to the provider

React Context would handle the notification part.

Reply:
That's a way of handling it, but it doesn't tell you what got updated, nor does it provide a way for components to clean up (`.off`) old listeners afaict, as doing it like this will just cause a general state update which will cause a re-`render()`?

Reply:
> nor does it provide a way for components to clean up (.off) old listeners afaict

Sure, which might be a sign of this not being a flexible enough solution